### DISCUSSION TITLE
Partial pre rendering

### DISCUSSION THREAD

**Discussion by asafnaory** [withastro/roadmap#852] [Proposal]
# Summary
Partial pre rendering  - a way to include SSG and SSR in the same page 

I assume you might have considered this idea since it's a feature present in Next.js. However, I haven't found any discussions or mentions of it anywhere

# Background & Motivation

As an Astro user, I aim to design pages that integrate Static Site Generation (SSG) and Server-Side Rendering (SSR). My goal is to have certain sections of the page statically generated, while others are rendered at runtime. For instance, consider a blog homepage that displays a list of available blog posts. I prefer not to have the entire page rendered at runtime since most of it is static. However, I also want to avoid fully static generation, as this might exclude newly added blog posts from being displayed (unless I use an island which is redundant since there is no interactivity in this page ). 

# Example

I believe the optimal approach to implement this feature is to extend the application of the `prerender` keyword, currently available only in pages, to `.astro` components more broadly. It would be beneficial if I could specify for each component whether I prefer it to be statically generated or rendered at runtime. Given that Astro already supports streaming, components rendered server-side (SSR) could be seamlessly integrated into the statically generated HTML.

components/Blogs.astro

```typescript
---
  const allBlogs = await blogService.getAllBlogs()  
  export const prerender = false; // SSR 
---
<BlogList allBlogs={allBlogs}>
```

pages/index.astro
```typescript
---
  export const prerender = true;  // SSG 
---
<h1>My Blog </h1>
<Blogs />  // <-- this component will be streamed to the client when it is ready 
```

**Comment by Matb85** [comment]
Hello, any updates on this? https://nextjs.org/docs/app/building-your-application/rendering/partial-prerendering

**Reply by delucis** [reply]
Yes, Astro now has “Server Islands” which supports static pages with server-rendered components.

Docs: https://docs.astro.build/en/guides/server-islands/

**Comment by ascorbic** [comment]
I'm going to close this, becauseit is now implemented as server islands:  https://docs.astro.build/en/guides/server-islands/
